crates/ruff_python_formatter/src/builders.rs

```diff
@@ -0,0 +1,216 @@
+use crate::context::NodeLevel;
+use crate::prelude::*;
+use crate::trivia::lines_before;
+use ruff_formatter::write;
+use rustpython_parser::ast::Ranged;
+
+/// Provides Python specific extensions to [`Formatter`].
+pub(crate) trait PyFormatterExtensions<'buf, 'context> {
```

Comment:
At rome, we only really had builder extensions but it might be useful to add some more (and move the trait?). For example, we could add a `contents()` or `comments` methods so that we don't need to always go through the context.

Reply:
I'll merge this to reduce the stack size. I can follow up with another PR if you want me to move it.